File: backend/internal/repository/user_pokemon_repo.go
package repository

import (
	"context"
	"fmt"
	"log/slog"
	"strconv"
	"time"

	"cloud.google.com/go/firestore"
	"github.com/kenyamamoto/pokelingual/backend/internal/model"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// UserPokemonRepo implements domain.UserPokemonRepository using Cloud Firestore.
type UserPokemonRepo struct {
	client *firestore.Client
}

// NewUserPokemonRepo creates a new UserPokemonRepo.
func NewUserPokemonRepo(client *firestore.Client) *UserPokemonRepo {
	return &UserPokemonRepo{client: client}
}

// UpsertEncounter creates or updates a user's encounter record for a Pokemon using a transaction.
func (r *UserPokemonRepo) UpsertEncounter(ctx context.Context, uid string, pokemonID int, score float64, captured bool) error {
	ref := r.client.Collection("users").Doc(uid).
		Collection("pokemon").Doc(strconv.Itoa(pokemonID))

	return r.client.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		doc, err := tx.Get(ref)
		if err != nil {
			if status.Code(err) == codes.NotFound {
				s := "seen"
				if captured {
					s = "captured"
				}
				now := time.Now()
				up := model.UserPokemon{
					PokemonID:         pokemonID,
					Status:            s,
					TotalCaptures:     boolToInt(captured),
					TotalEncounters:   1,
					LastEncounteredAt: now,
					BestScore:         score,
				}
				if captured {
					up.LastCapturedAt = now
				}
				return tx.Set(ref, up)
			}
			return fmt.Errorf("getting document: %w", err)
		}

		var existing model.UserPokemon
		if err := doc.DataTo(&existing); err != nil {
			return fmt.Errorf("parsing document: %w", err)
		}

		existing.TotalEncounters++
		existing.LastEncounteredAt = time.Now()

		if captured {
			existing.TotalCaptures++
			existing.Status = "captured"
			existing.LastCapturedAt = time.Now()
		}

		if score > existing.BestScore {
			existing.BestScore = score
		}

		return tx.Set(ref, existing)
	})
}

// GetCollection returns all discovered Pokemon for the given user, ordered by Pokemon ID.
func (r *UserPokemonRepo) GetCollection(ctx context.Context, uid string) ([]model.UserPokemon, error) {
	docs, err := r.client.Collection("users").Doc(uid).
		Collection("pokemon").
		OrderBy("pokemon_id", firestore.Asc).
		Documents(ctx).GetAll()
	if err != nil {
		return nil, fmt.Errorf("querying collection: %w", err)
	}

	result := make([]model.UserPokemon, 0, len(docs))
	for _, doc := range docs {
		var up model.UserPokemon
		if err := doc.DataTo(&up); err != nil {
			slog.Warn("failed to parse pokemon document, skipping", "doc_id", doc.Ref.ID, "error", err)
			continue
		}
		result = append(result, up)
	}

	return result, nil
}

// GetPokemon returns a specific Pokemon record for the given user.
func (r *UserPokemonRepo) GetPokemon(ctx context.Context, uid string, pokemonID int) (*model.UserPokemon, error) {
	doc, err := r.client.Collection("users").Doc(uid).
		Collection("pokemon").Doc(strconv.Itoa(pokemonID)).Get(ctx)
	if err != nil {
		return nil, fmt.Errorf("getting pokemon: %w", err)
	}

	var up model.UserPokemon
	if err := doc.DataTo(&up); err != nil {
		return nil, fmt.Errorf("parsing pokemon: %w", err)
	}

	return &up, nil
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

File: backend/src/repository/user-pokemon-repo.ts
import type { Firestore } from "@google-cloud/firestore";
import type { UserPokemonRepository } from "../domain/interfaces.js";
import type { UserPokemon } from "../types/index.js";

/** ユーザの図鑑進捗を Firestore に永続化する UserPokemonRepository 実装。 */
export class UserPokemonRepo implements UserPokemonRepository {
  private db: Firestore;

  constructor(db: Firestore) {
    this.db = db;
  }

  /** 遭遇/捕獲を 1 件記録する。既存ドキュメントがあれば集計値を更新する。 */
  async upsertEncounter(uid: string, pokemonID: number, score: number, captured: boolean): Promise<void> {
    const ref = this.db
      .collection("users").doc(uid)
      .collection("pokemon").doc(String(pokemonID));

    await this.db.runTransaction(async (tx) => {
      const doc = await tx.get(ref);

      if (!doc.exists) {
        const now = new Date();
        const data: UserPokemon = {
          pokemon_id: pokemonID,
          status: captured ? "captured" : "seen",
          total_captures: captured ? 1 : 0,
          total_encounters: 1,
          last_captured_at: captured ? now : null,
          last_encountered_at: now,
          best_score: score,
        };
        tx.set(ref, data);
        return;
      }

      const existing = doc.data() as UserPokemon;
      const now = new Date();

      existing.total_encounters++;
      existing.last_encountered_at = now;

      if (captured) {
        existing.total_captures++;
        existing.status = "captured";
        existing.last_captured_at = now;
      }

      if (score > existing.best_score) {
        existing.best_score = score;
      }

      tx.set(ref, existing);
    });
  }

  /** ユーザが遭遇したポケモン一覧をポケモンID昇順で返す。 */
  async getCollection(uid: string): Promise<UserPokemon[]> {
    const snapshot = await this.db
      .collection("users").doc(uid)
      .collection("pokemon")
      .orderBy("pokemon_id", "asc")
      .get();

    return snapshot.docs.map((doc) => doc.data() as UserPokemon);
  }

  /** 特定ポケモンのユーザ実績を取得する。未遭遇ならエラーを投げる。 */
  async getPokemon(uid: string, pokemonID: number): Promise<UserPokemon> {
    const doc = await this.db
      .collection("users").doc(uid)
      .collection("pokemon").doc(String(pokemonID))
      .get();

    if (!doc.exists) throw new Error(`pokemon not found: ${pokemonID}`);
    return doc.data() as UserPokemon;
  }
}
